### PR TITLE
Add numerical ordering option for string comparison operations

### LINKED ISSUE #13979: String comparer for sorting numeric strings logically
## Rationale

For sorting purposes it's common to need portions of strings containing numbers to be treated like numbers. Consider the list of strings `"Windows 7", "Windows 10"`.

Using the `Ordinal` `StringComparer` to sort the list one would get

```
Windows 10
Windows 7
```

but the desired ascending logical sort would be

```
Windows 7
Windows 10
```

## Proposed API

```diff
 namespace System {
     public class StringComparer {
+        public static StringComparer Create(CultureInfo culture, CompareOptions options);
     }
 }
 namespace System.Globalization {
     public enum CompareOptions {
+        NumericOrdering = 0x00000020
     }
 }
```

## Usage

```cs
var list = new List<string> { "Windows 10", "Windows 7" };
list.Sort(StringComparer.Logical); // List is now "Windows 7", "Windows 10"
```

This would also be good for sorting strings containing IP addresses.

## Details

* `Logical` is a convenience property equivalent to the result of `Create(CultureInfo.CurrentCulture, CompareOptions.Logical)`
* `LogicalIgnoreCase` is a convenience property equivalent to the result of `Create(CultureInfo.CurrentCulture, CompareOptions.Logical | CompareOptions.IgnoreCase)`
* Non-numeric sequences will be evaluated with the culture provided.
* Numeric sequences will be determined by the result of `Char.IsDigit`.
* All UTF-16 digits will be supported and are manually parsed using `Char.GetNumericValue`.
* Only positive integral values without digit separators will be supported directly.
* Numbers will be treated as `ulong`s. Logic for overflows will have to be considered.
* The string `Windows 8.1` would be considered 4 sequences. The `Windows ` would be a string sequence, the `8` would be a numeric sequence, the `.` would be another string sequence, and the `1` would be another numeric sequence.
* This API could later be expanded to include support for allowing signs, decimals, and digit separators through the use of overloads accepting a `NumberStyles` parameter.
* When a numeric and string sequence are considered at the same time the numeric sequence always comes before the string sequence so when sorting the following list, `"a", "7"` the number `7` will be sorted before the letter `a`.
* Existing methods that take a `CompareOptions` parameter as input will need to be updated to support the new `Logical` member.

## Open Questions

* Should `CompareOptions.Logical` be implemented as the flag option `SORT_DIGITSASNUMBERS` to the `dwCmpFlags` parameter of [`CompareStringEx`](https://msdn.microsoft.com/en-us/library/windows/desktop/dd317761(v=vs.85).aspx)? Using it's implementation should be more efficient but later expanding support for `NumberStyles` will require a re-implementation with matching behavior.

## Updates

* Added `Logical` and `LogicalIgnoreCase` properties.
* Added support for all UTF-16 digits.
* Added more `CreateLogical` overloads to match the `Create` method.
* Added retrieval of the `NumberFormatInfo` from the `StringComparer` parameter when not explicitly provided and is a `CultureAwareComparer`.
* Removed `CreateLogical` overloads that matched the `Create` method.
* Switched to only supporting positive integral values without digit separators.
* Added consideration of comparing a numeric sequence with a string sequence.
* Added the flag member `CompareOptions.Logical` and changed `CreateLogical` to be just an overload of `Create`.

### PR DESCRIPTION
**Pull request by PranavSenthilnathan**
Adds numerical ordering for comparison operations (e.g. Compare, Equals, GetHashCode, GetSortKey). This now enables comparisons of numbers based on their numerical value instead of lexicographical order, such as 2 < 10. We don't support Index operations (e.g. StartWith, EndsWith, IsPrefix, IsSuffix) since the underlying globalization libraries, NLS and ICU, don't support it.

Because this new option relies on underlying globalization libraries, there could be differences in behavior for different platforms and libraries:
- Full ICU (default in most cases):
  - Numbers with leading zeros are treated as equal to equivalent numbers with no leading zeros (`"01" == "1"`).
  - Equivalent numbers in different languages/scripts are treated the same (`"1" == "١"`, where ١ is the Arabic-Indic Digit One).
- NLS (legacy Windows):
  - Numbers with leading zeros are treated as unequal to equivalent numbers with no leading zeros (`"01" != "1"`). However, these numbers are compared as expected with unequal numbers, namely `1 < 02 < 2 < 03`.
  - Equivalent numbers in different languages/scripts are treated as unequal (`"1" != "١"`)
- Hybrid ICU:
  - WASM: The same as ICU, but the option set is limited. This PR does not change the valid option set besides allowing adding the NumericOrdering option to any previously valid option. [See this](https://github.com/dotnet/runtime/blob/f87428bdd39418a758d916bd950b82e3ee69835f/docs/design/features/globalization-hybrid-mode.md#wasm) for the allowed options.
  - iOS, tvos, maccatalyst: Not supported because I don't have an Apple device to test with. However, Apple does support numeric comparisons for strings.
    - [NSString.CompareOptions API](https://developer.apple.com/documentation/foundation/nsstring/compareoptions/1415530-numeric)
    - PAL change would probably be to recognize the dotnet CompareOptions and convert it to the NSString.CompareOptions [here](https://github.com/dotnet/runtime/blob/main/src/native/libs/System.Globalization.Native/pal_collation.m#L70)

Closes #13979.

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CompareOptions.cs
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/CompareOptions.cs:53]: >  /// </summary> [](http://example.com/codeflow?start=6&length=16)

will be good to add a remarks to this one giving more information like this option can be used in comparisons but not for search (IndexOf/StartsWith/EndsWith). Will be good to hint the behavior difference too when ICU is used against NLS. And last tell this option cannot be combined with the ordinal operations,

File: src/libraries/System.Runtime/ref/System.Runtime.cs
Comment (tarekgh) [inline src/libraries/System.Runtime/ref/System.Runtime.cs:9192]: nit: should we change this to use hex numbers for the sake of the readability?

File: src/libraries/System.Runtime/tests/System.Runtime.Tests/System/StringComparerTests.cs
Comment (tarekgh) [inline src/libraries/System.Runtime/tests/System.Runtime.Tests/System/StringComparerTests.cs:140]: Could you please test Equals method too here?

**Review by tarekgh — APPROVED** [review]
I left a few minor comments, LGTM otherwise. Thanks @PranavSenthilnathan!

I'll let @matouskozak and @ilonatommy comment on the hybrid and WASM stuff.

**Review by matouskozak — COMMENTED** [review]
Thank you for considering the Apple mobile globalization :). Could you please create an issue for the missing support on iOS/..., linking the original issue and the findings that you mentioned in the PR description. I'll look into it later.

I think we might have to temporarily disable the newly added tests for Apple mobile otherwise we will start getting `PlatformNotSupportedException` for the numeric ordering tests. I started the Apple mobile CI test runs to see if that's the case.


One more thing, in the PR description, there is an example for NLS
> However, these numbers are compared as expected with unequal numbers, namely `1 < 02 < 2 < 03`.

Does that mean that numbers with leading zeros are always smaller than without. Also, `002 < 02` then?

File: src/libraries/System.Runtime/ref/System.Runtime.cs
Comment (PranavSenthilnathan) [reply src/libraries/System.Runtime/ref/System.Runtime.cs:9192]: This is how the GetAPI tool generates it and the guidance I've heard is to make as few diffs from that tool as possible.

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CompareOptions.cs
Comment (PranavSenthilnathan) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CompareOptions.cs:54]: The comment was getting long so I just included just the part about indexing. Mentioning search might be confusing because people might consider IndexOf (which isn't supported) as search and GetHashCode (which is supported) as not. I think it's easier to just say that NumericOrdering works in all cases except for indexing.

I prefer to keep the combination behavior with Ordinal and OrdinalIgnoreCase on those members since this is really a property of theirs instead of numeric ordering. This is also consistent with the other options as well which don't mention Ordinal even though they can't combine with them either.

I think the ICU and NLS differences should probably go in docs rather than in doc comments since NLS usage is not going to be high. There are already docs about NLS/ICU differences that we can append to (https://learn.microsoft.com/en-us/dotnet/core/extensions/globalization-icu#behavioral-differences).

**Comment by PranavSenthilnathan** [comment]
> Thank you for considering the Apple mobile globalization :). Could you please create an issue for the missing support on iOS/..., linking the original issue and the findings that you mentioned in the PR description. I'll look into it later.

Created #109999

> I think we might have to temporarily disable the newly added tests for Apple mobile otherwise we will start getting `PlatformNotSupportedException` for the numeric ordering tests. I started the Apple mobile CI test runs to see if that's the case.

Updated the PR to skip them. I introduced a new PlatformDetection property IsNumericComparisonSupported instead of reusing the IsHybridGlobalizationOnApplePlatform so it's easier to find and remove once when you get the test working.

> One more thing, in the PR description, there is an example for NLS
> 
> > However, these numbers are compared as expected with unequal numbers, namely `1 < 02 < 2 < 03`.
> 
> Does that mean that numbers with leading zeros are always smaller than without. Also, `002 < 02` then?

Yes, if the numbers are actually equal, then in NLS the one with more leading zeros is considered less (it's basically a tiebreaker). This behavior is better for deterministic sorting of lists, but the downside is that hash tables won't consider these equal. I prefer ICU's behavior here (JS/wasm does the same) but I don't think there's much we can do about it.

File: src/libraries/System.Runtime/ref/System.Runtime.cs
Comment (tarekgh) [reply src/libraries/System.Runtime/ref/System.Runtime.cs:9192]: Yes, but still can make diffs :-)

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CompareOptions.cs
Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CompareOptions.cs:54]: Well, if you prefer that, it will be better to edit the docs of the indexing APIs and add the remark there. I am trying to make it easy for the API users to understand when this new enum value is not allowed. I guess users can be puzzled if they get exceptions and do not understand what is wrong. 

You don't have to block the PR on that, but it will be good to open a doc issue/PR to add the info as needed.

File: src/libraries/System.Runtime/ref/System.Runtime.cs
Comment (tannergooding) [reply src/libraries/System.Runtime/ref/System.Runtime.cs:9192]: The ref assembly isn’t really designed for readability, it’s designed to be correct and automatically generated

Deviations and manual diffs just cause later downstream pain and hinder the ability to rerun the tool, as people have to fight against it 

The better option would be to submit a bug or better a patch such that the output produced by the tool for flags enabled enums is “better” such as using hex or logical shifts to represent the bits instead (`1 << 0, 1 << 1, 1 << 2`, etc)

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CompareOptions.cs
Comment (PranavSenthilnathan) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CompareOptions.cs:54]: I like that idea. I just checked and we already list the valid options so I will update those docs once this goes in:
[Compare](https://learn.microsoft.com/en-us/dotnet/api/system.globalization.compareinfo.compare?view=net-8.0#system-globalization-compareinfo-compare(system-string-system-string-system-globalization-compareoptions))
[IndexOf](https://learn.microsoft.com/en-us/dotnet/api/system.globalization.compareinfo.indexof?view=net-8.0#system-globalization-compareinfo-indexof(system-string-system-char-system-int32-system-globalization-compareoptions))

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CompareInfo.Icu.cs
Comment (matouskozak) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/CompareInfo.Icu.cs:719]: I think this `#if` should be moved after the `if ((options & ValidCompareMaskOffFlags) != 0)` exception guard or the `AssertExtensions.Throws<ArgumentException>("options", () => ci.GetSortKey(s1, CompareOptions.Ordinal));` test from `SortKeyMiscTest` should get temporarily disabled for apple mobile.

File: src/libraries/System.Runtime/ref/System.Runtime.cs
Comment (PranavSenthilnathan) [reply src/libraries/System.Runtime/ref/System.Runtime.cs:9192]: I've created an issue: https://github.com/dotnet/sdk/issues/44999. GenAPI seems to be in a kind of limbo state where there is a new Roslyn based version we want to switch to (tracked by https://github.com/dotnet/sdk/issues/31088) and we don't want to maintain the current CCI-based one (https://github.com/dotnet/arcade/blob/main/src/Microsoft.DotNet.GenAPI/README.md).

**Review by matouskozak — APPROVED** [review]
Thank you for the test fixes. I checked the apple mobile CI and everything failing looks to be unrelated and tracked at https://github.com/dotnet/runtime/issues/103472. Looking good from Apple mobile side.

@ilonatommy do you want to check the WASM changes?

**Comment by ilonatommy** [comment]
> @ilonatommy do you want to check the WASM changes?

Sorry, I was away for a while. I read the WASM part and it looks good. Let me just run extra platforms with `_HybridGlobalization` suffix.

**Review by ilonatommy — APPROVED** [review]
`HybridGlobalization` failures are not connected.

**Pull request merged**
